### PR TITLE
Fix #15592

### LINKED ISSUE #15592: Generating a systemd file for a pod will put "--exit-policy" option after pod name, causing the service to fail
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

Running `podman generate systemd --new --name` against a pod without a manually defined `--exit-policy` will cause the option to be appended after the name, which causes the pod creation to fail.

**Steps to reproduce the issue:**

1. Create pod without specifying `--exit-policy`

2. Generate systemd unit with `podman generate systemd --new --name mypod`

**Describe the results you received:**

The following line in the unit file:

```
ExecStartPre=/usr/bin/podman pod create --infra-conmon-pidfile %t/pod-dns.pid --pod-id-file %t/pod-dns.pod-id --publish 53:53/tcp --publish 53:53/udp --publish 3000:3000/tcp --publish 853:853/tcp --publish 784:784/udp --publish 853:853/udp --publish 8853:8853/udp --publish 5443:5443/tcp --publish 5443:5443/udp dns --exit-policy=stop
```

**Describe the results you expected:**

The extra option is appended before the pod name.

**Additional information you deem important (e.g. issue happens only occasionally):**

This unit file can successfully start after moving "dns" to the end of the line. Manually running the command with minor modifications produces the following:

```
# /usr/bin/podman pod create --infra-conmon-pidfile /run/pod-dns.pid --pod-id-file /run/pod-dns.pod-id --publish 53:53/tcp --publish 53:53/udp --publish 3000:3000/tcp --publish 853:853/tcp --publish 784:784/udp --publish 853:853/udp --publish 8853:8853/udp --publish 5443:5443/tcp --publish 5443:5443/udp dns --exit-policy=stop
Error: accepts at most 1 arg(s), received 2
```

Manually specifying `--exit-policy` on the pod before generating the unit file avoids this issue.

**Output of `podman version`:**

```
podman version 4.2.0
```

**Output of `podman info`:**

```
host:
  arch: arm
  buildahVersion: 1.27.0
  cgroupControllers:
  - cpuset
  - cpu
  - io
  - memory
  - pids
  - rdma
  - misc
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.1.3-1.1.armv7hl
    path: /usr/bin/conmon
    version: 'conmon version 2.1.3, commit: unknown'
  cpuUtilization:
    idlePercent: 99.04
    systemPercent: 0.4
    userPercent: 0.55
  cpus: 4
  distribution:
    distribution: '"opensuse-microos"'
    version: "20220827"
  eventLogger: journald
  hostname: pi2-server
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 5.19.2-1-default
  linkmode: dynamic
  logDriver: journald
  memFree: 358948864
  memTotal: 998154240
  networkBackend: cni
  ociRuntime:
    name: runc
    package: runc-1.1.3-2.1.armv7hl
    path: /usr/bin/runc
    version: |-
      runc version 1.1.3
      commit: v1.1.3-0-ga916309fff0f
      spec: 1.0.2-dev
      go: go1.18.3
      libseccomp: 2.5.4
  os: linux
  remoteSocket:
    exists: true
    path: /run/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    seccompProfilePath: /etc/containers/seccomp.json
    selinuxEnabled: true
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.1.11-1.5.armv7hl
    version: |-
      slirp4netns version 1.1.11
      commit: unknown
      libslirp: 4.7.0
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.4
  swapFree: 0
  swapTotal: 0
  uptime: 20h 34m 23.00s (Approximately 0.83 days)
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries:
  search:
  - registry.opensuse.org
  - registry.suse.com
  - docker.io
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 4
    paused: 0
    running: 4
    stopped: 0
  graphDriverName: btrfs
  graphOptions: {}
  graphRoot: /var/lib/containers/storage
  graphRootAllocated: 15913713664
  graphRootUsed: 2584170496
  graphStatus:
    Build Version: Btrfs v5.18.1
    Library Version: "102"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 4
  runRoot: /run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 4.2.0
  Built: 1660176000
  BuiltTime: Thu Aug 11 00:00:00 2022
  GitCommit: ""
  GoVersion: go1.16.15
  Os: linux
  OsArch: linux/arm
  Version: 4.2.0
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-4.2.0-1.1.armv7hl
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/main/troubleshooting.md)**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**

A Raspberry Pi 2 device running OpenSUSE Micro OS

### PR DESCRIPTION
**Pull request by unit511**
Fix an issue that causes incorrect systemd unit files to
be generated if specified pod has been created without
specifying --exit-policy and named without --name.
The fix works by converting the pod creation command to
use --name if it isn't doing so already before it is passed
to setPodExitPolicy.

Closes #15592

Signed-off-by: Bence Göblyös <goblyos.bence@yahoo.com>

#### Does this PR introduce a user-facing change?

```
When using the `--new` flag, `podman generate systemd` will change the pod creation command to always use the `--name` flag.
```

### CODE REVIEW

**Comment by unit511** [comment]
#### Does this PR introduce a user-facing change?

```
When using the `--new` flag, `podman generate systemd` will change the pod creation command to always use the `--name` flag.
```

**Comment by mheon** [comment]
@vrothberg PTAL

File: pkg/systemd/generate/pods.go [hunk 1]
```diff
@@ -266,6 +266,16 @@ func setPodExitPolicy(cmd []string) []string {
...
 }
 
+// Make sure the name is specified by the --name argument
+func convertNameArg(cmd []string) []string {
+	for _, arg := range cmd {
+                if strings.HasPrefix(arg, "--name=") || arg == "--name" {
+                        return cmd
+                }
+        }
+	return append(cmd[:len(cmd)-1], "--name", cmd[len(cmd)-1])
```
Comment (vrothberg) [inline pkg/systemd/generate/pods.go:276]: What if we have `podman pod create`?  I think we need to pass the real name of the pod to this function and do the transformation only when the last index matches that name.

**Comment by vrothberg** [comment]
The linter isn't happy yet.  You can run it via `make validate`.

File: pkg/systemd/generate/pods.go (hunk 1 above)
Comment (unit511) [reply pkg/systemd/generate/pods.go:276]: Right, I didn't think of that. Checking the name of the pod is a good idea, I'll try to familiarize myself with the codebase a bit more and see what I can do. Some extra test cases for this scenario wouldn't hurt either.

**Comment by edsantiago** [comment]
@unit511 when you resubmit, please change the commit message to something that will help a future maintainer. Someone looking through `git log --oneline` would really like to see a brief description of the commit, not a random number. Thank you!

**Pull request closed without merging**